Dear ImGui Implementation

viniciuslrangel opened the pull request:
Press F10 to show FPS 😄 

- [x] Add core headers
- [x] Adapt SDL3 & Vulkan backend
- [x] Integrate to driver pipeline
- [x] Basic input handling (keyboard/mouse)
- [x] Gamepad capture
- [x] Layer abstraction
- [x] Basic video information layer (frame timing)

viniciuslrangel committed:
imgui renderer/basic input implementation

viniciuslrangel committed:
added imgui as dependency

sampletext32 commented on .gitmodules:
```diff
@@ -81,4 +81,9 @@
+	shallow = true
+[submodule "externals/dear_imgui"]
+	path = externals/dear_imgui
+	url = git@github.com:shadps4-emu/ext-imgui.git
```
all neighboring urls are `https`. 
```suggestion
	url = https://github.com/shadps4-emu/ext-imgui.git
```

abouvier commented on .gitmodules:
```diff
@@ -81,4 +81,9 @@
\ No newline at end of file
+	shallow = true
+[submodule "externals/dear_imgui"]
+	path = externals/dear_imgui
```
Why not just `imgui`?

viniciuslrangel replied:
I'll forward the declaration later

abouvier commented on externals/CMakeLists.txt:
```diff
@@ -149,6 +149,19 @@ if (APPLE AND NOT TARGET date::date-tz)
 endif()
 
+# Dear ImGui
+if (NOT TARGET Dear_ImGui)
```
The target will never exists if there is no `find_package` ;-)

viniciuslrangel committed:
imgui: add layers system

Add video info layer to show fps. Press F10 to toggle it.

viniciuslrangel committed:
imgui: add custom imgui config

viniciuslrangel committed:
imgui: gamepad capture, stopping propagation

viniciuslrangel committed:
imgui: changed config & log file path to use portable dir

viniciuslrangel committed:
videoout: render blank frame when video output is closed

required to render imgui even when game has no video output

- fixed merge compile-error

abouvier commented on externals/CMakeLists.txt:
```diff
@@ -155,6 +155,19 @@ if (APPLE)
 endif()
 
+# Dear ImGui
+block()
```
[block()](https://cmake.org/cmake/help/latest/command/block.html) was added in CMake 3.25 but shadps4 uses `cmake_minimum_required(VERSION 3.16.3)`.

viniciuslrangel replied:
Oh, thanks for the tip. Will fix later

raphaelthegreat reviewed (approved):
Looks awesome, thanks a lot for this work!
